newIDE/app/src/AssetStore/TagsHelper.js

```diff
@@ -0,0 +1,70 @@
+// @flow
+import { type AssetShortHeader } from '../Utils/GDevelopServices/Asset';
+
+const indexOfOccurrence = (
+  haystack: Array<string>,
+  needle: string,
+  occurrence: number
+) => {
+  var counter = 0;
+  var index = -1;
+  do {
+    index = haystack.indexOf(needle, index + 1);
+  } while (index !== -1 && ++counter < occurrence);
+  return index;
+};
+
+export const getFolderTagsFromAssetShortHeaders = ({
+  assetShortHeaders,
+  selectedFolders,
+}: {|
+  assetShortHeaders: Array<AssetShortHeader>,
+  selectedFolders: Array<string>,
+|}): Array<string> => {
+  /** @type {string[][]} */
+  const assetTagsAfterPackTag = [];
+  const allTagsAfterPackTag = new Set();
+  // We are in a pack, calculate first level folders based on asset tags.
+  // Tags are stored from top to bottom, in the list of tags of an asset.
+  // We first detect where the chosen category is, as this is the pack, and
+  // remove this tags and the others before (that could be bundles).
+
+  assetShortHeaders.forEach(assetShortHeader => {
+    const allAssetTags = assetShortHeader.tags;
+    const lastSelectedFolder = selectedFolders[selectedFolders.length - 1];
+    const occurencesOfLastSelectedFolderInSelectedFolders = selectedFolders.filter(
+      folder => folder === lastSelectedFolder
+    ).length;
+    const lastSelectedFolderIndex = indexOfOccurrence(
+      allAssetTags,
+      selectedFolders[selectedFolders.length - 1],
+      occurencesOfLastSelectedFolderInSelectedFolders
+    );
+    if (lastSelectedFolderIndex === -1) return allAssetTags; // This shouldn't happen, but just in case.
+    const tagsAfterPackTags = allAssetTags.filter(
+      (tag, index) => index > lastSelectedFolderIndex
+    );
+    if (tagsAfterPackTags.length > 0)
+      assetTagsAfterPackTag.push(tagsAfterPackTags);
+    tagsAfterPackTags.forEach(tag => allTagsAfterPackTag.add(tag));
+  });
+
+  // Then we remove the tags that are present in all assets, they're not useful, or not a folder.
+  // (For example: "pixel art")
```

Comment:
Does this code actually remove tags that are on every assets like the following one that is 2 time shorter?
It this code only an optimization?
If so, do you think it's worth the extra maintenance?
Is there a risk that the tags order change for some reason (a script that would use a Set at some point)?

Reply:
This first part ensures we remove all the top tags to only get the tags below the selected folder / pack.
It's needed to show a correct navigation.

It bases itself on the fact that the tags, generated by the script on the asset repository, are put in the right order, but apart from that I think we're fine and shouldn't have to touch this any more.

Reply:
I think it's worth noting in the GDevelop-assets repository somewhere that tags are supposed to be ordered so that "folders" can be re-constructed in the UI :)